zebrad/src/components/sync.rs

```diff
@@ -149,18 +129,31 @@ where
     ///  - verifier: the zebra-consensus verifier that checks the chain
     pub fn new(chain: Network, peers: ZN, state: ZS, verifier: ZV) -> Self {
         let tip_network = Timeout::new(peers.clone(), TIPS_RESPONSE_TIMEOUT);
-        let block_network = ServiceBuilder::new()
-            .retry(zn::RetryLimit::new(BLOCK_DOWNLOAD_RETRY_LIMIT))
-            .timeout(BLOCK_DOWNLOAD_TIMEOUT)
-            .service(peers);
-        let verifier = Timeout::new(verifier, BLOCK_VERIFY_TIMEOUT);
+        // The Hedge middleware is the outermost layer, hedging requests
+        // between two retry-wrapped networks.  The innermost timeout
+        // layer is relatively unimportant, because slow requests will
+        // probably be pre-emptively hedged.
+        //
+        // XXX add ServiceBuilder::hedge() so this becomes
+        // ServiceBuilder::new().hedge(...).retry(...)...
+        let block_network = Hedge::new(
+            ServiceBuilder::new()
+                .retry(zn::RetryLimit::new(BLOCK_DOWNLOAD_RETRY_LIMIT))
```

Comment:
I think the Retry should be outside the Hedge:
* with Hedge outside Retry, a slow request with a quick first retry gets tried 3 times: one for the initial slow request, once for the Retry on that request, and once by the quick Hedge retry
* with Retry outside Hedge, a slow request with a quick first retry gets tried 2 times: once for the initial slow request, and once by the quick Hedge retry

I think that's the way it works - but I might be missing some subtlety about the concurrency, selects and timeouts here.

If that's the case, let's explain why Hedge is outside Retry in the comments.

Reply:
The Hedge goes outside the Retry, because the Retry layer abstracts away spurious failures from individual peers making a less-fallible network service, and the Hedge layer tries to reduce latency of that less-fallible service.

Reply:
I added this comment in a suggestion.